src/Functional/Where.php

```diff
@@ -0,0 +1,81 @@
+<?php
+
+namespace Functional;
+
+use Functional\Exceptions\InvalidArgumentException;
+
+/**
+ * Return the subset of objects which exactly match the supplied properties
+ *
+ * @param iterable|Traversable|array $objects
+ * @param array $propertyMap
+ * @return array
+ */
+function where($objects, array $propertyMap)
+{
+    InvalidArgumentException::assertCollection($objects, __FUNCTION__, 1);
+
+    $result = [];
+    foreach ($objects as $obj)
+    {
+        if (matches($obj, $propertyMap)) {
+            $result[] = $obj;
+        }
+    }
+    return $result;
+}
+
+/**
+ * Helper function
+ *
+ * Check if an object or array (hashmap) contains the supplied values
+ *
+ * @param array|object $object
+ * @param array $propertyMap
+ * @return bool
+ */
+function matches($object, array $propertyMap)
+{
+    foreach ($propertyMap as $property => $value) {
+        if (is_array($value)) {
+            if (!matches(read_property($object, $property), $value)) {
+                return false;
+            } else {
+                continue;
+            }
+        }
+
+        // weak equality check to compare different instances of objects
+        if (read_property($object, $property) != $value) {
+            return false;
+        }
+    }
+    return true;
+}
+
+/**
+ * Helper function
+ *
+ * Read a property of an object with fallbacks for arrays and methods
+ *
+ * @param $object
+ * @param $property
+ * @return mixed
+ */
+function read_property($object, $property)
+{
+    if (is_object($object)) {
```

Comment:
I don't think there is a good way to determine if a magic property exist. The object should have implemented an __isset() method but there is no guarantee that it does and it might have different behavior for `null` which currently works fine.